sycl/include/sycl/esimd/esimd_math.hpp

```diff
@@ -0,0 +1,1950 @@
+          int flag = GENX_NOSAT) {
+  static_assert(SZ >= 8 && (SZ & 0x7) == 0,
+                "vector size must be a multiple of 8");
+
+  // __esimd_intrinsic_impl_pln() requires src1 and src2 to be combined into
+  // a single matrix, interleaving the values together in blocks of 8
+  // items (ie, a block of 8 from src1, then a block of 8 from src2, then
+  // the next block of 8 from src1, then the next block of 8 from src2,
+  // and so-on.)
+  simd<float, (SZ >> 3) * 16> Src12v;
+  auto Src12 = Src12v.template format<float, (SZ >> 3), 16>();
+
+  Src12.select<(SZ >> 3), 1, 8, 1>(0, 0) =
+      src1.template format<float, (SZ >> 3), 8>();
+  Src12.select<(SZ >> 3), 1, 8, 1>(0, 8) =
+      src2.template format<float, (SZ >> 3), 8>();
+
+  simd<float, SZ> Result = __esimd_pln<SZ>(src0, Src12.read());
+
+  if (flag != GENX_SAT)
+    return Result;
+
+  return esimd_sat<float>(Result);
+}
+
+// esimd_bf_reverse
+template <typename T0, typename T1, int SZ>
+ESIMD_NODEBUG ESIMD_INLINE simd<T0, SZ> esimd_bf_reverse(simd<T1, SZ> src0) {
+  simd<unsigned, SZ> Src0 = src0;
+  return __esimd_bfrev<unsigned>(Src0);
+}
+
+template <typename T0, typename T1>
+ESIMD_NODEBUG ESIMD_INLINE
+    typename std::enable_if<details::is_esimd_scalar<T0>::value &&
+                                details::is_esimd_scalar<T1>::value,
+                            typename std::remove_const<T0>::type>::type
+    esimd_bf_reverse(T1 src0) {
+  simd<T1, 1> Src0 = src0;
+  simd<T0, 1> Result = esimd_bf_reverse<T0>(Src0);
+  return Result[0];
+}
+
+// esimd_bf_insert
+template <typename T0, typename T1, int SZ, typename U, typename V, typename W>
+ESIMD_NODEBUG ESIMD_INLINE
+    typename std::enable_if<std::is_integral<T1>::value, simd<T0, SZ>>::type
+    esimd_bf_insert(U src0, V src1, W src2, simd<T1, SZ> src3) {
+  typedef typename details::dword_type<T1>::type DT1;
+  static_assert(std::is_integral<DT1>::value && sizeof(DT1) == sizeof(int),
+                "operand conversion failed");
+  simd<DT1, SZ> Src0 = src0;
+  simd<DT1, SZ> Src1 = src1;
+  simd<DT1, SZ> Src2 = src2;
+  simd<DT1, SZ> Src3 = src3;
+
+  return __esimd_bfins<DT1>(Src0, Src1, Src2, Src3);
+}
+
+template <typename T0, typename T1, typename T2, typename T3, typename T4>
+ESIMD_NODEBUG ESIMD_INLINE
+    typename std::enable_if<details::is_esimd_scalar<T0>::value &&
+                                details::is_esimd_scalar<T4>::value,
+                            typename std::remove_const<T0>::type>::type
+    esimd_bf_insert(T1 src0, T2 src1, T3 src2, T4 src3) {
+  simd<T4, 1> Src3 = src3;
+  simd<T0, 1> Result = esimd_bf_insert<T0>(src0, src1, src2, Src3);
+  return Result[0];
+}
+
+// esimd_bf_extract
+template <typename T0, typename T1, int SZ, typename U, typename V>
+ESIMD_NODEBUG ESIMD_INLINE
+    typename std::enable_if<std::is_integral<T1>::value, simd<T0, SZ>>::type
+    esimd_bf_extract(U src0, V src1, simd<T1, SZ> src2) {
+  typedef typename details::dword_type<T1>::type DT1;
```

Comment:
There are still a lot of `typedef` everywhere, but since you said it is preexisting code which is recycled it might make sense to keep them to avoid too much divergence between 2 code bases.

Reply:
So leave it as it is